src/test/lib/DataTest.cpp

```diff
@@ -1363,6 +1363,285 @@ QuicAbortiveTransfers(
+
+    uint8_t RawBuffer[] = "cancel on loss message";
+    QUIC_BUFFER MessageBuffer = { sizeof(RawBuffer), RawBuffer };
+
+    SelectiveLossHelper LossHelper; // used later to trigger packet drops
+
+    // Start the server.
+    MsQuicConfiguration ServerConfiguration(Registration, Alpn, Settings, ServerSelfSignedCredConfig);
+    TEST_TRUE(ServerConfiguration.IsValid());
+
+    CancelOnLossContext ServerContext{ DropPackets, true /* IsServer */, &ServerConfiguration};
+    QuicAddr ServerLocalAddr;
+
+    MsQuicListener Listener(Registration, CleanUpManual, QuicCancelOnLossListenerHandler, &ServerContext);
+    TEST_TRUE(Listener.IsValid());
+    TEST_EQUAL(Listener.Start(Alpn), QUIC_STATUS_SUCCESS);
+    TEST_EQUAL(Listener.GetLocalAddr(ServerLocalAddr), QUIC_STATUS_SUCCESS);
+
+    // Start the client.
+    MsQuicCredentialConfig ClientCredConfig;
+    MsQuicConfiguration ClientConfiguration(Registration, Alpn, Settings, ClientCredConfig);
+    TEST_TRUE(ClientConfiguration.IsValid());
+
+    CancelOnLossContext ClientContext{ DropPackets, false /* IsServer */, &ClientConfiguration};
+
+    // Initiate connection.
+    ClientContext.Connection = new MsQuicConnection(
+        Registration,
+        CleanUpManual,
+        QuicCancelOnLossConnectionHandler,
+        &ClientContext);
+    TEST_TRUE(ClientContext.Connection->IsValid());
+
+    QUIC_STATUS Status = ClientContext.Connection->Start(
+        ClientConfiguration,
+        QUIC_ADDRESS_FAMILY_INET,
+        QUIC_TEST_LOOPBACK_FOR_AF(QUIC_ADDRESS_FAMILY_INET),
+        ServerLocalAddr.GetPort());
+    if (QUIC_FAILED(Status)) {
+        TEST_FAILURE("Failed to start a connection from the client.");
+        return;
+    }
+
+    // Wait for connection to be established.
+    constexpr uint32_t EventWaitTimeoutMs{ 1'000 };
+
+    if (!ClientContext.ConnectedEvent.WaitTimeout(EventWaitTimeoutMs)) {
+        TEST_FAILURE("Client failed to get connected before timeout!");
+        return;
+    }
+    if (!ServerContext.ConnectedEvent.WaitTimeout(EventWaitTimeoutMs)) {
+        TEST_FAILURE("Server failed to get connected before timeout!");
+        return;
+    }
+
+    // Set up stream.
+    ClientContext.Stream = new MsQuicStream(
+        *ClientContext.Connection,
+        QUIC_STREAM_OPEN_FLAG_NONE,
+        CleanUpManual,
+        QuicCancelOnLossStreamHandler,
+        &ClientContext);
+    TEST_TRUE(ClientContext.Stream->IsValid());
+    Status = ClientContext.Stream->Start();
+    if (QUIC_FAILED(Status)) {
+        TEST_FAILURE("Client failed to start stream.");
+        return;
+    }
+
+    // Send test message.
+    Status = ClientContext.Stream->Send(&MessageBuffer, 1, QUIC_SEND_FLAG_CANCEL_ON_LOSS);
+    if (QUIC_FAILED(Status)) {
+        TEST_FAILURE("Client failed to send message.");
+        return;
+    }
+
+    // If requested, drop packets.
+    if (DropPackets) {
+        LossHelper.DropPackets(4); // does not work with 1 or 2, intermittently fails with 3
```

Comment:
Unlike the DatagramTest, this does not work when requesting only one packet to be dropped. Any ideas why that could be?

Reply:
See my comment above about PMTUD.